src/runtime_src/core/pcie/linux/system_linux.cpp

```diff
@@ -98,26 +110,76 @@ static std::string machine_info()
   return model;
 }
 
-static std::vector<std::weak_ptr<xrt_core::device_linux>> mgmtpf_devices(16); // fix size
-static std::vector<std::weak_ptr<xrt_core::device_linux>> userpf_devices(16); // fix size
-static std::map<xrt_core::device::handle_type, std::weak_ptr<xrt_core::device_linux>> userpf_device_map;
-
 }
 
 namespace xrt_core {
 
+void
+system_linux::
+register_driver(std::shared_ptr<pcidrv::pci_driver> driver)
+{
+  driver_list.push_back(driver);
+  if (driver->is_user())
+    driver->scan_devices(user_ready_list, user_nonready_list);
+  else
+    driver->scan_devices(mgmt_ready_list, mgmt_nonready_list);
+}
+
+std::shared_ptr<pcidev::pci_device>
+system_linux::
+get_pcidev(unsigned index, bool is_user) const
+{
+  if (is_user) {
+    if (index < user_ready_list.size())
+      return user_ready_list[index];
+    return user_nonready_list[index - user_ready_list.size()];
+  }
+
+  if (index < mgmt_ready_list.size())
+    return mgmt_ready_list[index];
+  return mgmt_nonready_list[index - mgmt_ready_list.size()];
+}
+
+size_t
+system_linux::
+get_num_dev_ready(bool is_user) const
+{
+  if (is_user)
+    return user_ready_list.size();
+  return mgmt_ready_list.size();
+}
+
+size_t
+system_linux::
+get_num_dev_total(bool is_user) const
+{
+  if (is_user)
+    return user_ready_list.size() + user_nonready_list.size();
+  return mgmt_ready_list.size() + mgmt_nonready_list.size();
+}
+
```

Comment:
Are all these list accesses thread safe?  Could a list be in process of being changed / updated by another thread?

Reply:
Duplicated comments?